irohad/torii/cache/abstract_cache.hpp

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_ABSTRACT_CACHE_HPP
+#define IROHA_ABSTRACT_CACHE_HPP
+
+#include <endpoint.pb.h>
+#include <boost/optional.hpp>
+#include <list>
+#include <mutex>
+#include <string>
+
+namespace iroha {
+  namespace cache {
+    /**
+     * Cache for any key-value types.
+     * Internally it uses a map to cache ItemTypes and linked list
+     * based index of KeyTypes to remove oldest items when getIndexSizeHigh() is
+     * reached. Implemented as a CRTP pattern.
+     * @tparam KeyType - type of cache keys
+     * @tparam ValueType - type of cache values
+     * @tparam T - need for CRTP
+     */
+
+    template <typename KeyType, typename ValueType, typename T>
+    class AbstractCache {
+     public:
+      /**
+       * @return high border of cache limit (@see AbstractCache#addItem)
+       */
+      uint32_t getIndexSizeHigh() const {
+        return static_cast<const T &>(*this).getIndexSizeHighImpl();
```

Comment:
There is no reason to use static_cast all time, better to write variable for it:
```
const T& impl = static_cast<const T &>(*this);
// usage
impl.getIndexSizeHighImpl();
```